src/util/mpscfifo.h

```diff
@@ -1,66 +1,88 @@
 #pragma once
 
 #include <QAtomicInt>
+#include <QMutexLocker>
 
 #include "util/assert.h"
+#include "util/memory.h"
 
-// Lock-free FIFO for multiple producers/writers and a single(!) consumer/reader.
+enum class MpscFifoConcurrency {
+    SingleProducer,
+    MultipleProducers,
+};
+
+// FIFO for multiple producers/writers and a single consumer/reader. Reading
+// is lock-free while concurrent writers are synchronized by a mutex. The
+// mutex can be disabled by explicitly creating a single producer instance.
 template<typename T, int capacity>
 class MpscFifo {
   public:
-    MpscFifo()
-          : m_enqueueSize(0),
-            m_dequeueSize(0),
-            m_headIndex(0),
-            m_tailIndex(0) {
-        static_assert(capacity >= 2, "capacity too low");
+    explicit MpscFifo(
+            MpscFifoConcurrency concurrency = MpscFifoConcurrency::MultipleProducers)
+          : m_writeTokens(0),
+            m_readTokens(capacity),
+            m_writeMutex((concurrency == MpscFifoConcurrency::MultipleProducers) ? new QMutex : nullptr),
```

Comment:
Can you move this to the body an use make_unique? I just stumbled about this, looking where memory is freed.

Reply:
I simply forgot to use `make_unique()` here :see_no_evil: